src/workerd/jsg/wrappable.c++

```diff
@@ -94,104 +192,49 @@ void Wrappable::maybeDeferDestruction(bool strong, kj::Own<void> ownSelf, Wrappa
   }
 }
 
-void Wrappable::traceFromV8(uint traceId) {
-  if (lastTraceId == traceId) {
-    // Duplicate trace, ignore.
-    //
-    // This can happen in particular if V8 choses to allocate an object unmarked but we determine
-    // that the object is already reachable. In that case we mark the object *and* run our own
-    // trace (because we can't be sure V8 didn't allocate the object already-marked), so we might
-    // get duplicate traces.
-  } else {
-    lastTraceId = traceId;
-    GcVisitor visitor(*this);
-    jsgVisitForGc(visitor);
-  }
+void Wrappable::traceFromV8(cppgc::Visitor& cppgcVisitor) {
+  cppgcVisitor.Trace(KJ_ASSERT_NONNULL(wrapper));
+  GcVisitor visitor(*this, cppgcVisitor);
+  jsgVisitForGc(visitor);
 }
 
 void Wrappable::attachWrapper(v8::Isolate* isolate,
                               v8::Local<v8::Object> object, bool needsGcTracing) {
   auto& tracer = HeapTracer::getTracer(isolate);
 
-  if (detached != NOT_DETACHED) {
-    // It appears that this Wrappable once had a wrapper attached, and then that wrapper was GC'd,
-    // but later on a wrapper was added again. This suggests a serious problem with our GC, in that
-    // it is collecting objects that are still reachable from JavaScript. However, we can usually
-    // continue operating even in the presence of such a bug: it'll only cause a real problem if
-    // a script has attached additional properites to the object in JavaScript and expects them
-    // to still be there later. This is relatively uncommon for scripts to do, though it does
-    // happen.
-#ifdef KJ_DEBUG
-    KJ_FAIL_ASSERT("Wrappable had wrapper collected and then re-added later");
-#else
-    // Don't crash in production. Also avoid spamming logs.
-    static bool alreadyWarned = false;
-    if (!alreadyWarned) {
-      kj::StringPtr collected;
-      switch (detached) {
-        case NOT_DETACHED:     collected = "NOT_DETACHED";     break;
-        case WHILE_SCAVENGING: collected = "WHILE_SCAVENGING"; break;
-        case WHILE_TRACING:    collected = "WHILE_TRACING";    break;
-        case OTHER:            collected = "OTHER";            break;
-      }
-      KJ_LOG(ERROR, "Wrappable had wrapper collected and then re-added later", collected,
-                    kj::getStackTrace(), lastTraceId, wrapper.getLastMarked(),
-                    detachedTraceId, tracer.currentTraceId());
-      alreadyWarned = true;
-    }
-#endif
-  }
+  KJ_REQUIRE(wrapper == nullptr);
+  KJ_REQUIRE(strongWrapper.IsEmpty());
 
-  KJ_REQUIRE(wrapper.IsEmpty());
-  wrapperRef = kj::addRef(*this);
-  wrapper.Reset(isolate, object);
+  auto& wrapperRef = wrapper.emplace(isolate, object);
```

Comment:
IIUC, this changes `Wrappable` such that a single instance *may* have multiple wrappers attached in its lifetime (even if unlikely), which would change the identity of that Wrappable as viewed in JavaScript -- that is, there's no longer a guarantee that the Wrappable is `===` equal when passed out to JS. Am I understanding correctly?

Either way, the following comment in `class MemoizedIdentity` in jsg.h should be updated:

```
Note that you don't need to wrap `jsg::Ref<T>` this way, as it already has the property that only one wrapper will be created.
```

Reply:
A `Wrappable` can now have multiple wrappers over its lifetime, but only one at a time. Before a second one can be created, the previous one would have to have been collected. This implies that it's impossible for someone to compare the old and new wrapper for identity equality because they cannot exist at the same time.

Reply:
Yep, got it.

Still, the comment there in MemoizedIdentity should be updated.

The one hypothetical case that *could* be a problem (isn't currently) is if someone later tries to take just the `GetIdentityHash()` of the wrapper now and maps that to the `Wrappable` instance.

Another *potentially* problematic case (but also somewhat hypothetical) is with WeakMaps. Let's suppose I use an API object as a key in a WeakMap but otherwise have no other JS reachable references, only a Ref somewhere in C++. 

```js
const wm = new WeakMap();
wm.set(foo.aWrappable, 123);

// ... gc scavenge runs and the foo.aWrappable wrapper is collected

// foo.aWrappable wrapper is recreated when accessed again...
wm.get(foo.aWrappable);  // undefined
```

Do we care? Does v8 treat api objects used as keys like this as unmodified and reclaimed in scavenge?

Reply:
They thought of that. :)

https://chromium.googlesource.com/v8/v8.git/+/refs/heads/main/src/objects/js-objects.cc#3019

Reply:
Nice.